### ISSUE TITLE
SpliceAI Built-in Filter

### ISSUE THREAD

**Issue by sksmi** [SACGF/variantgrid#216]
**Is your feature request related to a problem? Please describe.**
Requested by @PeerArts. Currently it's time-consuming to filter based on SpliceAI & most people don't know how to use it and/or aren't incorporating in their analysis. Would be good to add as a built-in filter at some point in the future as has been great in picking up deep intronic variants. 

**Describe the solution you'd like**
A built in filter where you can set a min threshold for all 4 SplicAI scores (and also other predictors?). 

**Describe alternatives you've considered**
Currently it's easy enough to make a manual splicing filter and merge this with variants from the damage node.

**Additional thoughts**
- Not sure if/how this interacts with the current 'keep splice variants checkbox'. 
- Needs validation prior to implementation to ensure we're not inadvertently excluding splice variants not covered by splicing predictors.
- Should filtering be part of the existing damage node or a separate 'splicing' node where you can set all splicing thresholds? 

@PeerArts any additional comments? Can you add a description of how you manage splicing filtering at the moment & any validation data you've used?

**Comment by davmlaw** [comment]
I think this is a dupe of "Analysis Splicing filters"- https://github.com/SACGF/variantgrid_private/issues/2442

**Comment by sksmi** [comment]
From 2442: 

SKS: Can we make a built in splicing filter? Potentially as part of the damage filter? Bit complex to add it at the moment. Currently working out appropriate thresholds, but looking at something like this:
Want to keep potential splice sites that don’t have a maxentscan or dbscSNV score (note maxent scan assesses larger splice region than dbscSNV, but still not nec. whole region of interest for research purposes) AND
[Splice sites with >threshold (?80%) maxentscan % diff OR
Sites >threshold (?0.6) dbscSNV(ada) OR
Sites >threshold (?0.6) dbscSNV (RF)]

DL: The way it currently works is there's a "always_keep_splice_variants_regardless_of_impact" option that returns anything that has any of the following:

consequence contains 'splice'
splice region is not null
dbscsnv_ada_score >= 0.6
dbscsnv_rf_score >= 0.6


Splicing Node - allow people to set min/max cutoffs etc etc for each of the different splicing algos

Eg for Splice AI cutoff Sarah likes 0.2

**Comment by sksmi** [comment]
TODO:
validate splicing filter thresholds & combos (SKS)
new splicing node w/ sashimi icon (DL)
summary/validation info on how research team approaches splicing (PA)

**Comment by PeerArts** [comment]
In some of my latest analyses I've tried to add a SpliceAI threshold of 0.2 for each of the 4, and added a filter for moderate and high impact variants (to avoid using the merge node). I haven't made a proper comparison/validation yet, but I think it may indicate possible splice events better, especially for deep(er) intronic and synonymous variants.
example here: https://variantgrid.com/analysis/2294/

![image](https://user-images.githubusercontent.com/34562740/106852216-9f4d3000-6707-11eb-9fe8-13d148acca15.png)

**Comment by davmlaw** [comment]
Do you think it's worth revisiting #146 - FilterNode - save / load - which could have been of use here to eg create someone's personal SpliceAI settings easily?

**Comment by sksmi** [comment]
No - I would still prioritise 100 other issues over filternode save/load.  It's pretty easy to manage manual filters using templates now Fx. I manage by having a couple of templates that fit most scenarios then just modify as required. My manual filters are a bit mad, but I constantly tweak them for different scenarios. It think it's far more important that we keep the built-in filters current and working well. 

![image.png](https://images.zenhubusercontent.com/5bc5465086f3a678d497c0be/575cc3a0-bc0b-4dba-99f4-ed3076ddd9e9)

**Comment by davmlaw** [comment]
I keep going back on whether to cram this into DamageNode or not, and it seems so do others (eg to avoid a merge)

The reason everyone avoids the merge of course is that it's slow - however in the case of a "diamond" pattern, we could make an optimisation which would make MergeNode as fast as if it was just 1 node, see #240

**Comment by davmlaw** [comment]
For SA Path - I think it's best to move this into damage for the performance - we can always move it out again later once MergeNode performance issues are done

I can do this pretty easily, and then you and KK can decide the details of thresholds later.

Will make v1 that has just 1 slider for all 4 SpliceAI values

**Comment by sksmi** [comment]
Still finalising splicing report, but 90% confident that the following should work:

Splice candidate defined as:
- splice region indels: [consequence contains 'splice' OR not null splice region] AND [variant class not SNV]
OR
- any splice predictor (RF,Ada or SpliceAI) >=splicing threshold (ST)

Prelim suggested ST:
- strict: 0.7
- mod 0.6
- permissive 0.2 or 0.4
Currently testing ROC characteristics to better define suggested thresholds, but think should be applied in node as slider as likely curators will want flexibility to change.
__________

If splicing implemented in damage node it makes more sense to me to treat whole node as an OR rather than AND. Esp. in the context of impact as currently if a revel or cadd score is applied when impact is set to mod, a bunch of high/mod impact variants are filtered out because they lack predictions. This is probably not what the majority of users want and it requires a second node + merge if you want to catch these variants.

Suggested change of impact text from:
Impact min xx (AND, but text is ambiguous and not documented)
to
'Keep variants with impact greater or equal to: xx' (OR, and same as splice variants)

**Comment by davmlaw** [comment]
OR issue covered in #121

Generally - I think it would be better to have simpler nodes, then allow people to combine them with Venn / Merges to build the logic they want. That being said, we're going to cram a lot into this node, so maybe an optional OR is good while it's rammed with options.

>> in the context of impact as currently if a revel or cadd score is applied when impact is set to mod, a bunch of high/mod impact variants are filtered out because they lack predictions

Not if you set allow null = True

I'm used to nulls, but maybe it would be better to have "require value" (default=False) on every filter than have the global "allow null"

There are also some conservation tracks we don't filter... we could add them too

**Comment by sksmi** [comment]
Agree re. simple as long as it's well documented. At the moment it's not clear to users how the filters on the damage node work & interact. 

Suggestion was to make it an OR rather than an AND for the whole node as that seems to make more sense given it will combine impact + prediction + splice, not so much add it as an option, although that might make more sense. 

> in the context of impact as currently if a revel or cadd score is applied when impact is set to mod, a bunch of high/mod impact variants are filtered out because they lack predictions

True, but even then you still lose some HIGH impact variants affecting stop/stop codons. Either way, it's not sufficiently clear that you need to all null in the majority of cases & I can't imagine many people not wanting to keep all HIGH variants in a default filter.

Conservation could be interesting although technically covered by REVEL & CADD (although so are polyphen etc).

Published and in domain would be of more interest to me as harder to make manual filter.

**Comment by sksmi** [comment]
My damage node brainstorm. If its possible to implement and useful I can run it past the user group for feedback. Text in green is suggested tooltip.

![image](https://user-images.githubusercontent.com/8601717/109248037-27bb7e00-7839-11eb-9803-2a3e9444eb81.png)

**Comment by sksmi** [comment]
@PeerArts any thoughts on the above?

**Comment by PeerArts** [comment]
Looks ok at first glance, happy to try after MRFF is due (Wednesday)

**Comment by davmlaw** [comment]
It looks to me that if you left at least 1 of the 6 sliders in the default position then it wouldn't filter anything.

Basically: OR means any is true, and they are set by default to allow through all variants, so unless you set them all to be non-defaults, the filter won't do anything.

>> you still lose some HIGH impact variants affecting stop/stop codons. .... I can't imagine many people not wanting to keep all HIGH variants in a default filter.

I think we def need an ALL/ANY ie logical AND/OR options - for backwards compatibility (legacy nodes would default to ALL)

I looked at an exome (900k variants) and 6305 were HIGH. I think people want to filter more stringently than that, and hence apply filters with an AND -  to make very strict filters when you're looking to find the low hanging fruit.

>> it's not sufficiently clear that you need to all null in the majority of cases

Quite a lot of the tools are only applied to a subset of variants, eg CADD/REVEL are missense only. Some are for SNVs only.

I think "Remove low complexity regions" (remove variants w/repeat masker) should always be applied - ie be outside of the AND/OR 

There's nothing in the UI about splicing that doesn't use the threshold, should they always be linked - ie there's no way to get only them but not the threshold filters, and when you set the threshold to non-zero it starts applying the other splicing filter?

**Comment by davmlaw** [comment]
>> you need to all null in the majority of cases

I think if you're using OR you don't want nulls (because you're looking for something to opt in)

When you're using AND, you want to be careful about nulls, because you don't want to throw out something that the tool didn't run over

**Comment by sksmi** [comment]
Most users don't apply the remove low complexity variants because they're scared something important will be lost.
Most users use the damage node to filter for all candidate variants, not the low hanging fruit. 
Yes the default position is confusing, but is the same as current settings - probably should be greyed out if not selected. Hiding/showing fields is confusing for users.

**Comment by davmlaw** [comment]
>> Most users don't apply the remove low complexity variants because they're scared something important will be lost.

I had a quick look and yeah nobody uses that.

Had a quick look at SA Path classificaitons, and low complexity would removes:

3 of 1439 Pathogenic (0.2%)
15 of 1091 B/LB (1.4%)

So, it kinda works, but yeah I agree probably dangerous/not worth it. Will remove that from filter.

>> Most users use the damage node to filter for all candidate variants, not the low hanging fruit.

This is a matter of degree, and depends how you define them. But here's a histogram of what percent of variants are in a damage node compared to a parent node (restricted to parents > 1k and < 100k)

![damage_histogram.png](https://images.zenhubusercontent.com/5bb45cdce135e46293defe80/52411036-bd4b-4cc6-a2a4-09f131e520ee)

Looks like the average DamageNode filters out ~90% of their parent

>> the default position is confusing, but is the same as current settings

The current settings apply them as an AND, which means that keeping them in the default position performs NO ADDITIONAL FILTER

If you change it to an OR, then leaving it in the default position means no FILTERING IN THE NODE AT ALL

Basically what this means is if you want the OR to work, you need a way to "not select" every single element

Another option is that instead of thinking of a group operation to apply to each filters, each filter could have an option:

* Not applied - N/A
* Include variants that match this filter - OR
* Must be met if present (ie nulls are ok) - AND (null)
* Required - AND

**Comment by sksmi** [comment]
old school opts to get around the default issue? 
![image.png](https://images.zenhubusercontent.com/5bc5465086f3a678d497c0be/d5b2e47c-9991-49dd-9177-62a9675f6604)  


![image.png](https://images.zenhubusercontent.com/5bc5465086f3a678d497c0be/3b9b8ce8-a7b2-4d1c-b7ca-1a689dada7b9)

**Comment by davmlaw** [comment]
Yeah looked into that - sliders don't have a null but you can make a fake lower bound, ie "-1" and then re-label that "N/A" and convert it on load/save.

**Comment by davmlaw** [comment]
Some test data to make sure migrations work ok: http://144.6.229.160/analysis/1965/

**BEFORE:**

32479 (leftmost node) - 6,527 variants

* impact min = HIGH
* always keep splice = True


32480 (node 2) - 8,249 variants

* impact min = MOD
* always keep splice = True
* CADD 20
* Revel 0.3
* allow null = True


32483 (node 3) - 7,435 variants

same as above but with null = false

* impact min = MOD
* always keep splice = True
* CADD 20
* Revel 0.3
* allow null = False

32481 (node 4) - 8,538 variants

* impact min = MOD
* always keep splice = True
* Damage 2 of 5

32478 (node 5) - 34,408 variants

* always keep splice = True
* Damage 2 of 5 

32482 (node 6) - 3,085 variants

* impact min = LOW
* Damage 2 of 5 

32484 (node 7) - 19,710 variants

* impact min = LOW
* Damage 2 of 5 
* allow null = True

**Comment by davmlaw** [comment]
Bugs found in testing:

* Removing enabled doesn't reset slider to NULL - ie save splicing value, set enabled = False, save. Doesn't save
* Need to deal with indels for damage node - probably benign is null should do the trick?

**Comment by davmlaw** [comment]
Something we need to document is that OR doesn't include null

For instance - if the only setting in the node is "2 of 5 damage predictors", it'll throw away anything that wasn't tested (ie indels, introns, intergenic etc etc)

If you only want that 1 filter, you can set it to required / null.

**Comment by sksmi** [comment]
Added text to readthedocs as think it's essential for users @davmlaw can you check for accuracy? Have made some assumptions - am basing testing off these. https://github.com/SACGF/variantgrid_docs/blob/master/analysis/nodes.md

**Comment by davmlaw** [comment]
Docs look good, thanks. I'm going to spend time buffing them up once everything's validated

Just wanted to make sure everything was good before writing help so I didn't have to rewrite it

**Comment by sksmi** [comment]
Closing as deployed to prod. 
Feature not tested.
